[dnf5]: comps: Support for environments

pkratoch opened the pull request:
Requires: https://github.com/openSUSE/libsolv/pull/468

pkratoch committed:
comps: Support for environments

pkratoch committed:
comps: Test that group/environment contains only relevant solvables

lukash commented on include/libdnf/comps/environment/query.hpp:
```diff
@@ -0,0 +1,121 @@
+namespace libdnf {
+
+class Base;
+using BaseWeakPtr = WeakPtr<Base, false>;
```
Please use the definition in `include/libdnf/base/base_weak.hpp` :slightly_smiling_face:

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,277 @@
+
+std::string Environment::get_environmentid() const {
+    std::string solvable_name(
+        lookup_str(query->sack->comps.p_impl->get_pool(), environment_ids, SOLVABLE_NAME));
```
This is a common occurrence across the codebase. You store the `query` here and only use it to get to the `Pool` in a really complicated way. I'm working on a PR to move the `Pool` to `Base` and I think we should only store `Base` (a `BaseWeakPtr` in particular) in all these objects, unless there's a reason to store something else. Saying this as a heads-up but also if you feel there could be an issue that we should discuss.

pkratoch replied:
I don't see any issue with this. There is also a question if we want a separate pool for comps, but I don't know, I am just letting you know there was a discussion around this.

lukash commented on include/libdnf/comps/environment/environment.hpp:
```diff
@@ -0,0 +1,139 @@
+};
+
+
+inline EnvironmentId::EnvironmentId(int id) : id(id) {}
```
Could you put the definition inline into the class declaration? It just saves on the number of characters :slightly_smiling_face:

lukash commented on libdnf-cli/output/groupinfo.hpp:
```diff
@@ -127,4 +127,4 @@ void print_groupinfo_table(libdnf::comps::Group & group) {
 }  // namespace libdnf::cli::output
 
-#endif  // LIBDNF_CLI_OUTPUT_REPOLIST_HPP
+#endif  // LIBDNF_CLI_OUTPUT_GROUPINFO_HPP
```
This would be nicer in a separate commit but feel free to keep it (same for ` libdnf-cli/output/grouplist.hpp`).

pkratoch replied:
I added one commit for all the minor fixes in groups. Is it fine that way?

lukash replied:
Sure, thanks!

lukash commented on libdnf/comps/environment/environment-private.hpp:
```diff
@@ -0,0 +1,47 @@
+/*
```
We use underscores, not dashes in filenames, this could use fixing for the `group-private.hpp` file too.

pkratoch replied:
The `group-private.hpp` and `environment-private.hpp` files were removed altogether.

lukash commented on libdnf/comps/environment/sack_impl.hpp:
```diff
@@ -0,0 +1,44 @@
+class EnvironmentSack::Impl {
+public:
+    explicit Impl();
+    ~Impl();
```
AFAICS you don't need the constructor and destructor at all (the definition in the .cpp is an empty body). FWIW by convention we don't create a `*_impl.cpp`, we put the definitions into the same `.cpp` file as the main class definitions (e.g. here it would go into `sack.cpp`).

lukash commented on libdnf/comps/environment/sack_impl.hpp:
```diff
@@ -0,0 +1,44 @@
+    ~Impl();
+
+private:
+    WeakPtrGuard<EnvironmentSack, false> sack_guard;
```
Also, if the only thing you need to hold in the pimpl is this `WeakPtrGuard`, I think you can just put that into the public header and drop the pimpl altogether. It's better to not have it if possible (@jrohel agreed?).

pkratoch replied:
I removed the `Impl` for both sack and query.

lukash commented on include/libdnf/comps/environment/query.hpp:
```diff
@@ -0,0 +1,117 @@
+    EnvironmentQueryWeakPtr get_weak_ptr();
+
+protected:
+    EnvironmentQuery(EnvironmentQuery && query);
```
You declare a move constructor here, but there's no definition in the corresponding `query.cpp`, there's a copy constructor instead. On the topic of copy/move constructors in general please see my commit where I've cleaned this up in the advisory classes: https://github.com/rpm-software-management/libdnf/commit/6e3b186abdecec3264087da03f7d89f18280b853 (the commit description gives some explanation, feel free to ask if something's unclear).

Also, what's the reasoning behind this one being protected?

pkratoch replied:
I cleaned up the constructors. I believe the Environment and EnvironmentQuery don't need any of the copy and move stuff now (the groups are the same), since I removed the weak pointer from Environment and therefore also the WeakPtrGuard in EnvironmentQuery, and so the default constructors can be used.

I'm still not sure about EnvironmentSack and Comps though. Right now, they can't be copied/moved, because their WeakPtrGuard can't be. But I don't know if they even should be copied/moved.

lukash commented on libdnf/comps/environment/query_impl.hpp:
```diff
@@ -0,0 +1,44 @@
+    ~Impl() = default;
+
+private:
+    WeakPtrGuard<EnvironmentQuery, false> data_guard;
```
Same as `Sack`, pimpl is unnecessary if it just contains the guard.

lukash commented on libdnf/comps/environment/sack.cpp:
```diff
@@ -0,0 +1,33 @@
+EnvironmentSack::EnvironmentSack(Comps & comps) : Sack(), comps{comps}, p_impl{new Impl()} {}
+
+
+EnvironmentSack::~EnvironmentSack() {}
```
Please use `= default` instead of an empty body definition.

lukash commented on libdnf/comps/environment/sack.cpp:
```diff
@@ -0,0 +1,33 @@
+#include <solv/solvable.h>
+}
+
+#include <map>
```
This seems like way too many includes, most of them aren't needed here.

lukash commented on libdnf/comps/environment/environment-private.hpp:
```diff
@@ -0,0 +1,47 @@
+namespace libdnf::comps {
+
+
+void add_solvable_id(Environment & environment, Id solvable_id);
```
This function isn't used anywhere, remove it?

lukash commented on libdnf/comps/environment/environment-private.hpp:
```diff
@@ -0,0 +1,47 @@
+void add_solvable_id(Environment & environment, Id solvable_id);
+
+
+void add_solvable_ids(Environment & environment, std::vector<Id> solvable_ids);
```
This one is only used in one place, you can move it to just that `.cpp` file?

lukash commented on libdnf/comps/environment/environment-private.hpp:
```diff
@@ -0,0 +1,47 @@
+
+// Search solvables that correspond to the environment_ids for given key
+// Return first non-empty string
+std::string lookup_str(Pool * pool, std::vector<EnvironmentId> environment_ids, Id key);
```
Have a look at `libdnf::solv::Pool::lookup_str()`, should do what you need.

lukash replied:
(I haven't said it explicitly but it seems you can drop this header altogether, doesn't seem to be needed)

pkratoch replied:
The ``libdnf::solv::Pool::lookup_str()`` returns the string only for one id, but I need to search over all the environment ids.

lukash replied:
Okay, sorry haven't noticed. You could still use the `libdnf::solv::Pool::lookup_str()` in a loop, and since you need that in 3 places in `environment.cpp`, you could put that into a local function. I see the same method is also duplicated in `comps/group.cpp`, so perhaps we should share that function more globally?

I'm not entirely sure what's the situation, are the `environment_ids` multiple libsolv `Id`s for environments with the same string id from multiple repos? And are they somehow guaranteed to e.g. have the same name or could it happen you could get different names depending on which one is first in the `environment_ids` vector?

pkratoch replied:
> Okay, sorry haven't noticed. You could still use the `libdnf::solv::Pool::lookup_str()` in a loop, and since you need that in 3 places in `environment.cpp`, you could put that into a local function. I see the same method is also duplicated in `comps/group.cpp`, so perhaps we should share that function more globally?

It might make sense, but I am not sure how to do it when groups have ``GroupId`` objects and environments have ``EnvironmentId`` and there is no duck-typing. But it's both just for storing solvable id, so maybe it can be united.

> I'm not entirely sure what's the situation, are the `environment_ids` multiple libsolv `Id`s for environments with the same string id from multiple repos? And are they somehow guaranteed to e.g. have the same name or could it happen you could get different names depending on which one is first in the `environment_ids` vector?

Yes, they are for environments with the same string id. They have to be merged and it can be different for each query (because different repos can be enabled). There were two options:
1. Create one merged environment object when query is created
2. Store only the solvable ids and iterate through them when merged attribute is needed

Originally, we've decided to use the option 2 because I hadn't found a way to iterate over all translations and could only return translation when language was given. This has changed (it's used in ``libdnf::comps::Environment::dump``), but the option 2 stayed.

What is still missing is to order the solvables, because it hasn't been yet decided how it should be ordered.

lukash replied:
> It might make sense, but I am not sure how to do it when groups have `GroupId` objects and environments have `EnvironmentId` and there is no duck-typing. But it's both just for storing solvable id, so maybe it can be united.

Oh, I see. Okay just leave it this way, but you don't need the `environment-private.hpp` to share the function with other modules, right?

And we can leave the merging topic for later, but thanks for the explanation :slightly_smiling_face:

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,297 @@
+#include <libdnf/base/base.hpp>
```
Missing the copyright / license header.

lukash reviewed (commented):
I've decided to send an incomplete review, looks like the code needs some cleanups :slightly_smiling_face: A lot of the comments apply to same issues in multiple places, is it ok to just comment on one occurrence and would you go through the files and fix them all?

Also, semi-related, I think a lot of this applies to the already-merged comps group code, but that might be for another PR to unify and clean that up too.

pkratoch commented on include/libdnf/comps/group/group.hpp:
```diff
@@ -131,10 +125,10 @@ class Group {
 
 private:
-    GroupQueryWeakPtr query;
+    libdnf::BaseWeakPtr base;
```
I went with weak pointer to base, but I am not sure if it's the right choice, because I can also imagine comps or sack could be used.

lukash commented on include/libdnf/comps/group/query.hpp:
```diff
@@ -75,8 +73,7 @@ class GroupQuery : public libdnf::sack::Query<Group> {
 
-    class Impl;
-    std::unique_ptr<Impl> p_impl;
+    WeakPtrGuard<GroupQuery, false> data_guard;
```
I'm quite sure we don't want weak pointers with a guard on queries (they are created on-demand by the API user, as opposed to owned by base and accessed through a getter via a reference, or in our case, the weak pointer).

Hence I think you can drop this guard, as well as `get_weak_ptr()`, the copy constructor (which creates a new instance of `data_guard` in the copy, thus circumventing the guard's purpose, and for completeness, you're missing an assignment operator that'd go with it :slightly_smiling_face: ) and the destructor.

pkratoch replied:
I dropped the query weak pointers in the last commit, so this change wouldn't have any effect at the end. But I can do it before that if you think the commits would make more sense that way.

lukash commented on libdnf/comps/group/query.cpp:
```diff
@@ -36,10 +35,11 @@ namespace libdnf::comps {
 
-GroupQuery::GroupQuery(const GroupSackWeakPtr & sack) : Query(), sack(sack), p_impl{new Impl()} {
+
+GroupQuery::GroupQuery(const GroupSackWeakPtr & sack) : Query(), sack(sack) {
```
You don't need to explicitly call the base class default constructor, you can just leave it out:
```c++
GroupQuery::GroupQuery(const GroupSackWeakPtr & sack) : sack(sack) {
```

lukash commented on libdnf/comps/group/sack.cpp:
```diff
@@ -40,11 +39,11 @@ namespace libdnf::comps {
 
 
-GroupSack::GroupSack(Comps & comps) : Sack(), comps{comps}, p_impl{new Impl()} {}
+GroupSack::GroupSack(Comps & comps) : Sack(), comps{comps} {}
```
Again, no need to call `Sack()`.

lukash commented on libdnf/comps/group/sack.cpp:
```diff
@@ -40,11 +39,11 @@ namespace libdnf::comps {
+GroupSack::GroupSack(Comps & comps) : Sack(), comps{comps} {}
 
 
 GroupSack::~GroupSack() {}
```
I don't think you need the destructor anymore (as there's no longer a `p_impl`), you can delete it.

pkratoch replied:
When I try to delete it, there is en `error: invalid application of ‘sizeof’ to incomplete type ‘libdnf::comps::Group’`

lukash commented on include/libdnf/comps/comps.hpp:
```diff
@@ -17,15 +17,14 @@ You should have received a copy of the GNU Lesser General Public License
 along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 */
 
+
```
I think we only use one linebreak in this place (between copyright notice and includes), though we don't have this written anywhere...

pkratoch replied:
I thought it's the other way around, because there are double linebreaks in some places. I changed it.

lukash commented on include/libdnf/comps/group/group.hpp:
```diff
@@ -31,10 +32,11 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 public:
     GroupId() = default;
-    explicit GroupId(int id);
+    explicit GroupId(int id) : id(id){};
```
Space before curled braces, not sure why clang format didn't catch this:
```c++
    explicit GroupId(int id) : id(id) {};
```

pkratoch replied:
clang-format doesn't think so :(

lukash commented on include/libdnf/comps/group/package.hpp:
```diff
@@ -35,16 +37,16 @@ enum class PackageType { MANDATORY, DEFAULT, OPTIONAL, CONDITIONAL };
 class Package {
 public:
     explicit Package(std::string name, PackageType type, std::string condition)
```
Any non-basic type arguments (here the `std::string`s) should under normal circumstances always be passed by a const reference for the least amount of overhead. Then below, instead of moving, you just do a copy via `name(name)`. That's the canonical way to do this in C++.

```c++
   explicit Package(const std::string & name, PackageType type, const std::string & condition)
```

For completeness you could pass all arguments this way, the argument there is that a simple type is something up to 8B in size (on 64bit archs), so you're not saving anything by using a const reference which is of the same (pointer) size.

pkratoch replied:
Thanks, fixed.

lukash reviewed (commented):
@pkratoch could you please rebase? Also, if you'd post the first 5 commits (everything before the big one with environments) separately, we could get that in quickly and then handle the env code.

pkratoch commented:
> @pkratoch could you please rebase? Also, if you'd post the first 5 commits (everything before the big one with environments) separately, we could get that in quickly and then handle the env code.

Thanks, I rebased it and made a separate PR for the first 5 commits: https://github.com/rpm-software-management/libdnf/pull/1416

lukash commented on include/libdnf/comps/environment/environment.hpp:
```diff
@@ -0,0 +1,128 @@
+    void add_environment_id(EnvironmentId environment_id) { environment_ids.push_back(environment_id); }
+
+    /// Get environment id
+    std::string get_environmentid() const;
```
This is intentionally without an underscore (as opposed to `get_environment_id()`) to distinguish between the ids of the `EnvironmentId` type and the string id, right? If so it's not a great interface because the two are so similar and it appears to be an inconsistency at first glance.

On a related note, more of a question for discussion, why are we exposing the libsolv ids (returned from `get_environment_ids()`) on the interface?

pkratoch replied:
I think both the methods that expose the libsolv ids can be removed (`get_environment_ids` and `add_environment_id`). It is used only in `EnvironmentQuery`, but the query can access the `environment_ids` directly. It would also make the interface better, even though there would still be the private `environment_ids` variable. Does it sound good?

lukash replied:
I think it does, though I'm not sure if the IDs would be needed (or just useful to have) on the API. The decision to have them or not on the API should be based on that, not on us having a naming problem :slightly_smiling_face:

pkratoch replied:
The naming problem was just a side note. It makes sense to me to not have them on the interface. And it would be easier to add them later when we find out it's needed than vice versa.

lukash replied:
Ok, let's drop them from API for now and we'll see.

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,317 @@
+std::string Environment::get_environmentid() const {
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    std::string solvable_name(lookup_str(pool, environment_ids, SOLVABLE_NAME));
```
When you only use a variable once, I prefer to just put the value in its place directly (unless it's too long), you save up on SLOCs and IMO it's usually also more readable (the variable is an extra link you need to make):
```c++
std::string solvable_name(lookup_str(get_pool(base), environment_ids, SOLVABLE_NAME));
```

This repeats quite a lot of times in methods below.

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,317 @@
+std::string Environment::get_translated_name(const char * lang) const {
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    // Go through all environment solvables and return first translation found.
+    for (EnvironmentId environment_id : environment_ids) {
+        Solvable * solvable = pool.id2solvable(environment_id.id);
+        if (const char * translation = solvable_lookup_str_lang(solvable, SOLVABLE_SUMMARY, lang, 1)) {
+            // Return translation only if it's different from the untranslated string
+            // (solvable_lookup_str_lang returns the untranslated string if there is no translation).
+            const char * untranslated = solvable_lookup_str(solvable, SOLVABLE_SUMMARY);
+            if (translation != untranslated && strcmp(translation, untranslated) != 0) {
+                return std::string(translation);
+            }
+        }
+    }
```
This block of code is repeated 3 more times below and then 4 more times in `group.cpp`. It could actually be refactored into a method, couldn't it?
```c++
std::string lookup_translated_str(libdnf::solv::Pool & pool, std::vector<EnvironmentId> environment_ids, Id key, const char * lang = nullptr) {
    for (EnvironmentId environment_id : environment_ids) {
        Solvable * solvable = pool.id2solvable(environment_id.id);
        const char * translation = nullptr;
        if (lang) {
            translation = solvable_lookup_str_lang(solvable, key, lang, 1);
        } else {
            translation = solvable_lookup_str_poollang(solvable, key);
        }
        if (translation) {
            // Return translation only if it's different from the untranslated string
            // (solvable_lookup_str_lang returns the untranslated string if there is no translation).
            const char * untranslated = solvable_lookup_str(solvable, key);
            if (translation != untranslated && strcmp(translation, untranslated) != 0) {
                return std::string(translation);
            }
        }
    }
}
```

There'd be an issue with sharing it with the goup code in that `GroupId` and `EnvironmentId` are different types, which can be solved by templating the function.

Another issue is if the translation isn't found (not handled in my example code above). I'm not sure why you actually compare the translation with the untranslated string, since at the end of each function you seem to return the untranslated string anyway? Granted it's retrieved a bit differently from libsolv, but don't they actually result in the same string being returned? IIRC I've once checked and libsolv's `pool_lookup_str()` and `solvable_lookup_str()` return the same thing, barring some special cases which I don't think apply here.

Then this and the `lookup_str()` function above could be shared in a file (possibly called `libdnf/comps/pool_utils.{hpp,cpp}` or similar). I recall we discussed sharing the `lookup_str()` right? That one small function maybe wasn't worth it but if there were two and this second one a bit more complex, seems sharing the code would be better).

pkratoch replied:
> There'd be an issue with sharing it with the goup code in that `GroupId` and `EnvironmentId` are different types, which can be solved by templating the function.

Oh, cool, I'll do that then.

> Another issue is if the translation isn't found (not handled in my example code above). I'm not sure why you actually compare the translation with the untranslated string, since at the end of each function you seem to return the untranslated string anyway? Granted it's retrieved a bit differently from libsolv, but don't they actually result in the same string being returned? IIRC I've once checked and libsolv's `pool_lookup_str()` and `solvable_lookup_str()` return the same thing, barring some special cases which I don't think apply here.

I compare the translation with the untranslated string, because libsolv returns the untranslated string if the translation doesn't exist. And I have to check every time, because theoretically there could be different untranslated string for every solvable. I could get rid of exactly one search (the one at the end) by remembering the first one, but I think it would be even more confusing.

> Then this and the `lookup_str()` function above could be shared in a file (possibly called `libdnf/comps/pool_utils.{hpp,cpp}` or similar). I recall we discussed sharing the `lookup_str()` right? That one small function maybe wasn't worth it but if there were two and this second one a bit more complex, seems sharing the code would be better).

Ok, thanks.

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,317 @@
+            std::string group_solvable_name = pool.id2str(*r_id);
+            groups.push_back(group_solvable_name.substr(group_solvable_name.find(":") + 1, std::string::npos));
+        }
+    }
```
This piece of code is also repeated below and could be shared. Basically you'd do:
```c++
if (groups.empty() {
    groups = load_group_from_pool(get_pool(base), solvable->dep_requires);
}

return groups;
```

as the body of this function and then the rest will be in the new `load_group_from_pool()`, passing the initial value for `r_id` as an argument.

pkratoch replied:
I'm not sure I understood correctly, but the code is now shared.

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,317 @@
+        result.emplace(solvable->repo->name);
+    }
+    return result;
+}
```
I'm not sure what's the intention for this method, in general wouldn't it be better to return a set (or rather a vector, for that matter) of the repo objects themselves, or rather weak pointers to them? A pointer to the Repo object should be in `solvable->repo->appdata` (you'd need to static_cast it). No need to resolve this one in the PR.

pkratoch replied:
I added just a TODO note for now.

lukash commented on include/libdnf/comps/environment/environment.hpp:
```diff
@@ -0,0 +1,128 @@
+    const std::vector<EnvironmentId> & get_environment_ids() const { return environment_ids; }
+    void add_environment_id(EnvironmentId environment_id) { environment_ids.push_back(environment_id); }
+
+    /// Get environment id
```
I take it the docstrings aren't in their final form, but when you add them you can try to make them closer to what they should be, in this case (for the simple getters) I'd do:
```c++
    /// @return The environment id.
```

(I think we should consistently stick to using full sentences (start with a capital letter and end with a full stop), for the cases where more than  one sentence is needed)

Applies to all the docstrings in the commit.

lukash commented on include/libdnf/comps/environment/environment.hpp:
```diff
@@ -0,0 +1,128 @@
+    /// Get translated description of a environment based on current locales.
+    /// If a translation is not found, return untranslated description.
+    ///
+    /// @replaces dnf:dnf/comps.py:attribute:Environment.ui_description
```
Please use only two slashes for the `@replaces`, so that it isn't a part of the docstring.

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,317 @@
+
+bool Environment::get_installed() const {
+    auto repos = get_repos();
+    return repos.find("@System") != repos.end();
```
```c++
return get_repos().find("@System") != repos.end();
```
Seems strictily better to me :slightly_smiling_face:

pkratoch replied:
In most cases I agree, but here it doesn't seem better, because I'd have to call get_repos() twice.
`return get_repos().find("@System") != get_repos().end();`

lukash replied:
Oh, you're right in this case, of course.

lukash commented on libdnf/comps/environment/environment.cpp:
```diff
@@ -0,0 +1,317 @@
+
+    // Save the document
+    if (xmlSaveFormatFileEnc(path.c_str(), doc, "utf-8", 1) == -1) {
+        throw std::runtime_error("failed to save xml document for comps");
```
Please don't throw `std::runtime_error` from libdnf code. We have custom exception hierarchy. I'm not entirely sure how to approach this as I don't know what kinds of errors can happen here besides filesystem errors.

Is there any error-handling mechanism in libxml that would provide an actual error message to include in the exception? I did some changes to the `xmlSaveFormatFileEnc()` so I think I would have checked for it (I don't remember), so there likely isn't... The next question would be could we use a better library that actually does error handling...

If this was only about filesystem errors, in other places we use `std::filesystem::filesystem_error` (because this is thrown from `std::filesystem` routines and to have a single exception class for the filesystem-related errors we throw it from our code as well, when it fits).

Otherwise please define your own exception class, inheriting from `libdnf::Error`. It could either be something like `EnvironmentError` (a generic error related to envs), or in this case maybe xml-specific `XmlSaveError` (which you'd share with `Group::dump()`).

pkratoch replied:
I haven't found any error-handling mechanism in libxml. The saving can fail for multiple reasons, returning `-1` in all the cases, not just filesystem errors, so I added the `XmlSaveError` as you suggested.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -0,0 +1,97 @@
+    for (Id solvable_id : solvable_ids) {
+        environment.add_environment_id(EnvironmentId(solvable_id));
+    }
+}
```
This method is only used once and is very short, it would be better to put the code in-place.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -0,0 +1,97 @@
+        }
+        // Map environmentids with list of corresponding solvable_ids
+        // TODO(pkratoch): Sort solvable_ids for each environmentid according to something (repo priority / repo id / ?)
+        environmentid = solvable_name.substr(delimiter_position, std::string::npos);
```
The "parsing" of the `solvable_name`, i.e. the split on `":"` should be factored out into a function. In other places you do a `+ 1` on the delimiter, not here, it just ends up not mattering as it's just used as a key in the map.

The way you do it (copying the libsolv's `char *` to `std::string solvable_name` and then making another substring from that) is also not very efficient. I think we could use `std::string_view` for this, using the libsolv's original `char *` as the stored string and only making views of that. This could be done with the other strings pulled from libsolv (descriptions etc.) too. We can discuss off-pr.

pkratoch replied:
Oh, I forgot to check for other places where `std::string_view` could be used. I'll check it out.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -0,0 +1,97 @@
+            environmentid.append("_available");
+        } else {
+            environmentid.append("_installed");
+        }
```
I don't like this appending of suffixes. This is just for the distinction in the aggregating map, you can instead make the map key a pair:
```c++
std::map<std::pair<std::string, bool>, std::vector<Id>> environment_map;
```

The `bool` represents the installed/available.

But, can there even be more than one installed environment with the same `environmentid`? In theory there shouldn't, right? In that case you could `add()` the installed envs directly without aggregating them in the map.

Could you also document this mechanism in comments (i.e. that it aggregates the envs based on their name and also distinguishes installed and available)? The comment you have now (_"Map environmentids with list of corresponding solvable_ids"_) is quite generic and one still needs to study the code to understand.

pkratoch replied:
Yeah, there shouldn't be more than one installed environment. I changed it so that installed environments are added right away and the map then contains only the available ones, so it's bit simpler now.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -0,0 +1,97 @@
+        if (environment_map.find(environmentid) == environment_map.end()) {
+            std::vector<Id> solvable_ids;
+            environment_map.emplace(environmentid, solvable_ids);
+        }
```
AFAIK you don't need to do this at all. The `environment_map[environmentid]` on the line below will create an empty vector as the value in the map if not yet present and then return a reference to it.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -0,0 +1,97 @@
+            std::vector<Id> solvable_ids;
+            environment_map.emplace(environmentid, solvable_ids);
+        }
+        environment_map[environmentid].insert(environment_map[environmentid].begin(), solvable_id);
```
What's the reason for inserting at the beginning? That'd be useful to have in a comment.

pkratoch replied:
There is a related TODO note and after the changes, it's right above this line, so it should be clear now.

lukash commented on include/libdnf/comps/environment/query.hpp:
```diff
@@ -0,0 +1,100 @@
+class EnvironmentQuery : public libdnf::sack::Query<Environment> {
+public:
+    // Create new query with newly composed environments (using only solvables from currently enabled repositories)
+    explicit EnvironmentQuery(const EnvironmentSackWeakPtr & sack);
```
Could you remve the sack constructor and just use the base one? Some time ago we agreed everything should be constructible from base to keep things simple and so having a different way to construct here just adds unnecessary variance.

lukash commented on include/libdnf/comps/environment/sack.hpp:
```diff
@@ -0,0 +1,73 @@
+
+class EnvironmentSack : public libdnf::sack::Sack<Environment> {
+public:
+    ~EnvironmentSack();
```
I don't think you need this explicit destructor declaration (and definition in cpp) anymore.

FWIW not much being left in the sack, but I'd keep it as it is for now before we decide what to do about the sacks.

pkratoch replied:
It's still needed since there is the `WeakPtrGuard` and it's needed for `get_group_sack` in `Comps`. Well, at least until we decide what to do about the sacks.

lukash commented on microdnf/commands/environment/environment.cpp:
```diff
@@ -0,0 +1,55 @@
+
+void EnvironmentCommand::run() {
+    auto & cmd = *get_argument_parser_command();
+    cmd.help();
```
Again, why not just do:
```c++
get_argument_parser_command()->help();
```

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    // Create environments based on the environment_map
+    for (auto it = environment_map.begin(); it != environment_map.end(); it++) {
```
You can iterate the map using the ranged `for` in a bit more straightforward way:
```c++
for (const auto & item : environment_map) {
    // ...
}
```
The `item` is a `std::pair` [key, value].

lukash commented on microdnf/commands/environment/environment_info.cpp:
```diff
@@ -0,0 +1,106 @@
+        // --available / all -> add available not-installed environments into the list
+        auto query_available = libdnf::comps::EnvironmentQuery(query);
+        query_available.filter_installed(false);
+        std::set<std::string> installed_environmentids;
+        for (auto environment : query_installed.list()) {
+            installed_environmentids.insert(environment.get_environmentid());
+        }
```
You're filling the `installed_environmentids` but never use it, unless I'm missing something you can remove this. Same in `environment_list.cpp`.

pkratoch replied:
Well, this one is partially on you because of https://github.com/rpm-software-management/libdnf/commit/a32851aec170add068a83ebd0d44b733ec4ce767 :P

lukash commented on microdnf/commands/environment/environment_info.cpp:
```diff
@@ -0,0 +1,106 @@
+        for (auto environment : query_available.list()) {
+            environment_list.emplace(environment);
+        }
+    }
```
I'm not sure I see the point of the block above adding the installed and available sets to the output separately. I'd think it's to order the installed first, but you're adding them all to a `std::set`, so the order is defined by that. Why not just leave the query unfiltered by installed / available instead?

Also the same in `environment_list.cpp`. In fact the bodies of `run()` are way too similar between these two. Let's see if we can whittle this down to something more trivial, if not, I'd prefer to share the code again in some way.

pkratoch replied:
It was done in this way to hide the duplicates, but it was changed in https://github.com/rpm-software-management/libdnf/commit/a32851aec170add068a83ebd0d44b733ec4ce767 . I was not aware of it, but I'll take a look now.

lukash commented on test/data/repos-repomd/repomd-comps-core-environment/repodata/repomd.xml:
```diff
@@ -0,0 +1,18 @@
+    <timestamp>1641802880</timestamp>
+    <size>492</size>
+  </data>
+</repomd>
```
As a bit of a side-note I think it's the `update-checksums.py` script that doesn't add a trailing newline to the files.

I don't require it for this PR but we should fix it eventually so we don't have these inconsistencies.

lukash commented on test/libdnf/comps/test_environment.hpp:
```diff
@@ -0,0 +1,51 @@
+    void test_merge_with_empty();
+    void test_merge_empty_with_nonempty();
+    void test_merge_different_translations();
+    void test_dump();
```
Could you also add some tests for `EnvironmentQuery`, basically testing all the filters and at least one test that has more than one result?

lukash reviewed (commented):
Apologies for the longish review, hope it makes sense :slightly_smiling_face: 

Could you also try to rebase? I think I broke your code again when moving the `load_repos()` from microdnf to libdnf, sorry about that...

pkratoch committed:
comps: Don't expose libsolv ids of groups on the interface

pkratoch committed:
comps: Minor fixes - remove unnecessary variables, fix includes

pkratoch committed:
comps: Add fuction to split solvable name into a type:id pair

pkratoch committed:
comps: Improve query costructors to create map only for available groups

Installed groups and environments should have only one solvable each,
so they can be added right away.

pkratoch committed:
comps: Simplify filtering in group list and info commands

The old code was originally used to aviod duplicates (when one group
was both installed and available), but it was partially removed.

pkratoch committed:
comps: Replace GroupSackWeakPtr with BaseWeakPtr in GroupQuery

pkratoch committed:
comps: Add XMLSaveError to be used when failed to save xml document

pkratoch committed:
comps: Fix docstrings, add missing and make @replaces only a comment

pkratoch committed:
comps: Add a TODO item for the get_repos method

pkratoch committed:
comps: Add equality operators for groups and change conparison

pkratoch committed:
comps: Add tests for GroupQuery filtering

pkratoch commented on test/libdnf/utils.hpp:
```diff
@@ -93,6 +97,70 @@ struct assertion_traits<libdnf::advisory::AdvisorySet> {
+    inline static std::string toString(const libdnf::comps::Environment & environment) {
+        std::string repos;
+        for (const auto & repo : environment.get_repos()) {
+            if (!repos.empty()) {
+                repos += ", ";
+            }
+            repos += repo;
```
This is not great, but not sure how to do it better, without the `if` part. Also, I assume this method is just to uniquely describe an Environment, so that's why I am adding the id and repos to the string.

lukash replied:
I think this is just fine. Yes, the purpose is to print all important data of the object in the unittest output. You can add more than the necessary unique identifier if you think it'd be useful and won't add too much clutter. In this case the id and the list of repos seems good.

pkratoch commented on test/libdnf/utils.hpp:
```diff
@@ -155,6 +223,8 @@ struct assertion_traits<libdnf::base::TransactionPackage> {
 
 
 std::vector<libdnf::advisory::Advisory> to_vector(const libdnf::advisory::AdvisorySet & advisory_set);
+std::vector<libdnf::comps::Environment> to_vector(const libdnf::Set<libdnf::comps::Environment> & environment_set);
+std::vector<libdnf::comps::Group> to_vector(const libdnf::Set<libdnf::comps::Group> & group_set);
```
I use `libdnf::Set` for sets of `Groups`/`Environments`, but I don't know if it's correct - if it shouldn't rather be `std::set` (but then it couldn't be easily converted from `GroupQuery`) or the query class (that would be probably too restrictive?), and why there is for example `AdvisorySet` as a special class instead of having just `std::set<Advisory>` or `libdnf::Set<Advisory>`.

lukash replied:
Using `libdnf::Set` is correct at this stage of things AFAIU. The reason `AdvisorySet` is a special class is because it uses `SolvMap` as its underlying storage (same as PackageSet), which is more effective at least for when you need to pass the `SolvMap` to libsolv (the fact that `SolvMap` is a bitmap has other advantages and also some disadvantages). Your items aren't represented by a single solvable Id so you can't use that.

As for why not just use `std::set` (which is actually the underlying storage for `libdnf::Set`), it's just to keep a consistent interface with the `SolvMap`-based sets. I'm not sure how important is that and maybe we should consider if we should just use `std::set` for simplicity. Not sure atm.

pkratoch committed:
comps: Add function get_translated_str to handle all comps translations

lukash commented on libdnf/comps/group/query.cpp:
```diff
@@ -42,43 +42,45 @@ namespace libdnf::comps {
-            group_map.emplace(groupid, solvable_ids);
+            // Create map of available groups:
+            // for each groupid (SOLVABLE_NAME), list all corresponding solvable_ids
+            // TODO(pkratoch): Sort solvable_ids for each groupid according to repo id
```
I wonder if we should resolve this TODO in the PR. It seems semi-important and might be good to do while you've got everything fresh in the head. Let's discuss it tomorrow.

pkratoch replied:
You can find the sorting in the last but one commit now.

lukash reviewed (commented):
Great job @pkratoch and nicely put together, thanks! Just one thing below I think we might wanna do now.

pkratoch committed:
comps: Sort group_ids and environment_ids by the source repoid

Also fix tests according to the change and add tests where the repos
are loaded in different order, but the groups/environment are merged
in the same way (still according to the repoids).

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -43,10 +44,11 @@ namespace libdnf::comps {
     Solvable * solvable;
-    std::pair<std::string, std::string> solvable_name_pair;
+    std::pair<std::string_view, std::string> solvable_name_pair;
+    const char * repoid;
```
You could declare `repoid` as `std::string_view` directly.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
+
         // Add installed environments directly, because there is only one solvable for each
-        if (strcmp(solvable->repo->name, "@System") == 0) {
+        if (strcmp(repoid, "@System") == 0) {
```
If `repoid` was `std::string_view`, you could do `repoid == "@System"` here, which is a bit nicer in C++.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
-            // TODO(pkratoch): Sort solvable_ids for each environmentid according to repo id
             available_map[solvable_name_pair.second].insert(
-                available_map[solvable_name_pair.second].begin(), solvable_id);
+                available_map[solvable_name_pair.second].end(), std::pair<Id, std::string_view>(solvable_id, repoid));
```
... and here you could use `std::make_pair()`, which is a bit less verbose way to construct a pair (you don't need to specify the template arguments, they are inferred). If `repoid` was `std::string_view` :slightly_smiling_face:.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
 
     // Create environments based on the available_map
-    for (const auto & item : available_map) {
+    for (auto item : available_map) {
```
If you don't use a reference, this makes a copy of each item in the map. `auto & item` is much more effective (if you don't mind modifying the data in the map).

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
-        for (Id solvable_id : item.second) {
-            environment.environment_ids.push_back(EnvironmentId(solvable_id));
+        // Sort the vector of (solvable_id, repoid) pairs by repoid
+        std::sort(item.second.begin(), item.second.end(), [](auto a, auto b) { return a.second > b.second; });
```
Since you're storing the repoid in the vector now (which makes sense, I originally thought you'd pull the repoid from the solvables during the comparison but this is much more effective), if you put the repoid as the first item in the vector, you wouldn't need the lambda at all, it would work automatically via operators on the pair and string_view.

With the lambda, you should again use `auto &` (or `const auto &`), otherwise you're making copies just for the comparison.

pkratoch replied:
It still needs to be reversed, so that the behavior is the same as in dnf 4, so I used `std::greater<>()`. Maybe I could instead iterate it in reverse afterwords, but I don't know which method is more efficient.

lukash replied:
Isn't it less intuitive that way? Do we need to preserve compatibility with this? I'd just do the "natural" ordering unless I'm missing something...

pkratoch replied:
When getting for example group name, the `group_ids` are searched in order and first that is found is used. So, if ordered naturally, it would take name from repo `fedora` instead of `updates`. I don't know how big problem it would cause if we changed the order. Probably the comps don't change that often, but I'd rather keep it as it is.

lukash replied:
Ok, I see. Do we have this documented somewhere in the comments? (sorry I don't remember and don't want to go through everything again)

I think we should have this documented in some appropriate place in the code in detail, including the "coincidental" interaction with the `fedora` and `updates` repos, this is an important information that's not obvious.

pkratoch replied:
Good idea, I added it to the declaration of group_ids and environment_ids.

lukash replied:
Perfect, thank you :slightly_smiling_face:

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
+        // Sort the vector of (solvable_id, repoid) pairs by repoid
+        std::sort(item.second.begin(), item.second.end(), [](auto a, auto b) { return a.second > b.second; });
+        // Create environment_ids vector from the sorted solvable_ids
+        for (auto solvableid_repoid_pair : item.second) {
```
Again, use reference, here it can be `const auto &`.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
+        std::sort(item.second.begin(), item.second.end(), [](auto a, auto b) { return a.second > b.second; });
+        // Create environment_ids vector from the sorted solvable_ids
+        for (auto solvableid_repoid_pair : item.second) {
+            environment.environment_ids.push_back(EnvironmentId(solvableid_repoid_pair.first));
```
You could do `environment.environment_ids.emplace_back(solvableid_repoid_pair.first);` to construct the vector items in-place, which should be a bit faster and is less verbose.

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -43,10 +44,11 @@ namespace libdnf::comps {
     libdnf::solv::Pool & pool = get_pool(base);
 
-    std::map<std::string, std::vector<Id>> available_map;
+    std::map<std::string, std::vector<std::pair<Id, std::string_view>>> available_map;
```
It might be helpful to document what's what in the container structure, the code is a bit hard to read if one doesn't know.

lukash reviewed (commented):
On a side note I've been re-running the CI jobs, it's randomly failing and I don't know why :confused:

lukash commented on libdnf/comps/environment/query.cpp:
```diff
@@ -43,10 +44,15 @@ namespace libdnf::comps {
+    // Map of available environments:
+    //     For each environmentid (SOLVABLE_NAME) have a vector of (repoid, solvable_id) pairs.
+    //     Each pair consists of one solvable_id that represents one definition of the environment
+    //     and repoid of its originating repository (in the opposite order).
```
Nitpicking, but _"in **reverse** order"_, though the sorting is done explicitly after the map is filled up so I'd leave out the note here.

pkratoch replied:
In this case, I just meant the order within the pair, because I first had to mention solvable_id and then repoid, but the pair is (repoid, solvable_id). But I see it's confusing, so I'll remove it. And it's mentioned before, so it will be fine without it.
